File: exercise/scala/day13/src/main/scala/santamarket/model/Receipt.scala
package santamarket.model

import scala.collection.mutable.ListBuffer

case class Receipt():
  private val items = ListBuffer[ReceiptItem]()
  private val discounts = ListBuffer[Discount]()

  def getTotalPrice: Double =
    items.map(_.totalPrice).sum + discounts.map(_.discountAmount).sum

  def addProduct(product: Product, quantity: Double, price: Double, totalPrice: Double): Unit =
    items += ReceiptItem(product, quantity, price, totalPrice)

  def getItems: List[ReceiptItem] = items.toList

  def addDiscount(discount: Discount): Unit =
    discounts += discount

  def getDiscounts: List[Discount] = discounts.toList

  override def equals(obj: Any): Boolean = obj match
    case that: Receipt =>
      this.items == that.items && this.discounts == that.discounts
    case _ => false

  override def hashCode(): Int = (items, discounts).hashCode

File: exercise/kotlin/day13/src/main/kotlin/santamarket/model/Receipt.kt
package santamarket.model

class Receipt {
    private val items = mutableListOf<ReceiptItem>()
    private val discounts = mutableListOf<Discount>()

    fun getTotalPrice(): Double {
        var total = 0.0
        items.forEach { item -> total += item.totalPrice }
        discounts.forEach { discount -> total += discount.discountAmount }
        return total
    }

    fun addProduct(product: Product, quantity: Double, price: Double, totalPrice: Double) {
        items.add(ReceiptItem(product, quantity, price, totalPrice))
    }

    fun getItems(): List<ReceiptItem> = items.toList()

    fun addDiscount(discount: Discount) {
        discounts.add(discount)
    }

    fun getDiscounts(): List<Discount> = discounts.toList()

    override fun equals(other: Any?): Boolean =
        other is Receipt && items == other.items && discounts == other.discounts

    override fun hashCode(): Int = 31 * items.hashCode() + discounts.hashCode()
}
